### PR TITLE
First release of the JS client

### PR DESCRIPTION
**Pull request by pboothe**
Had to give up on the dream of typescript.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/ndt7-js/2)

### CODE REVIEW

File: src/ndt7-upload-worker.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+   * @param {*} data
+   * @param {*} start
+   * @param {*} end
+   * @param {*} previous
+   * @param {*} total
+   */
+  function uploader(data, start, end, previous, total) {
+    if (closed) {
+      // socket.send() with too much buffering causes socket.close(). We only
+      // observed this behaviour with pre-Chromium Edge.
```
Comment (mattcat10) [inline src/ndt7-upload-worker.js:78]: Are you noticing this in safari as well? I have ran the code locally and see this happening in Safari Version 13.1.1 (15609.2.9.1.2).

Comment (bassosimone) [reply src/ndt7-upload-worker.js:78]: Does https://bassosimone.github.io/ndt7/ work for you with Safari? I just checked and it seems to work for me when using Safari from mobile and from Desktop, but I only have 20 Mbit/s in upload. If you confirm that it works, then maybe we can see what are the differences in the two implementation and apply fixes. (All implementations of ndt7 in JavaScript that I know of are derived from the same small client, but there are some differences now.)

Comment (mattcat10) [reply src/ndt7-upload-worker.js:78]: The test in that link seems to work without any socket errors, though it seems to only send part of the upload test, it seems the test is timing out early, the max packets i am seeing is 2.1MB in size

Comment (bassosimone) [reply src/ndt7-upload-worker.js:78]: What does the bottom of the console looks like? Mine is like this for a test run using Safari 13.1.2:

```
[Log] {AppInfo: {ElapsedTime: 8643000, NumBytes: 20257344}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] {AppInfo: {ElapsedTime: 8904000, NumBytes: 21305881}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] {AppInfo: {ElapsedTime: 9156000, NumBytes: 21305856}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] {AppInfo: {ElapsedTime: 9407000, NumBytes: 21305856}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] ndt test complete – {Origin: "client", Test: "upload", WorkerInfo: {ElapsedTime: 10003000, Error: "Terminated with timeout"}} (ndt7, line 216)
[Log] ndt: complete (ndt7, line 123)
```

Thank you for checking!

Comment (andrewRichardson) [reply src/ndt7-upload-worker.js:78]: I work on the same team as Matt and have faster internet (300 mbps upload) and this is what I'm seeing on Safari 13.1.2:

Screen Shot 2020-08-17 at 2 01 26 PM

Here is the tail of the console as well:

image

Comment (bassosimone) [reply src/ndt7-upload-worker.js:78]: Thanks a bunch! That is very useful to know! Maybe I can try using Docker and localhost to see if I can reproduce!

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
Adding comments to checkpoint progress on this PR and merge code as-is.

> Reviewed 3 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/m-lab/ndt7-js/2#-:-MIzpmbUCYCWrfON5NcE:b-1vpw4k)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mattcat10, @pboothe, and @stephen-soltesz)

___
*[README.md, line 2 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu25bN2-5H4u7AHBuU:-MIu25bN2-5H4u7AHBuV:b-md55xr) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/README.md#L2)):*
> ```Markdown
> The official [NDT7](https://github.com/m-lab/ndt-server) Javascript client
> libraries. This code that works both with Node.js and on the web and is the
> ```

nit: word choice.
___
*[README.md, line 5 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu2C7RCKQGDaIEhkKf:-MIu2C7RCKQGDaIEhkKg:b-f4hdg5) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/README.md#L5)):*
> ```Markdown
> source for the npm package `ndt7`.
> 
> Includes an example web client and an example node.js client in the `src/`
> ```

nit: in the examples/ dir.
___
*[examples/node-client.js, line 22 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu3oAl7PmOOi7-rF0h:-MIu3oAmBqSo1GS4J4cr:b-nav14v) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/examples/node-client.js#L22)):*
> ```JavaScript
>       },
>       downloadComplete: function(data) {
>         // (bytes/second) * (bits/byte) * (megabits/bit) = Mbps
> ```

nit: comment is out of sync with logic.
___
*[examples/node-client.js, line 33 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu4aGfE6KwhaKaJtyT:-MIu4aGfE6KwhaKaJtyU:b58h7lx) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/examples/node-client.js#L33)):*
> ```JavaScript
>         // bytes * (bits/byte() * (megabits/bit) * (1/seconds) = Mbps
>         const serverBw =
>             data.LastServerMeasurement.TCPInfo.BytesReceived * 8 / 1000000 / 10;
> ```

TODO: Upload time (10sec) should be dynamic.
___
*[src/ndt7.js, line 93 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu5NgM1k4V_UclvaDL:-MIu5NgNBnWGuOKRKDAm:b-lrf7a4) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/src/ndt7.js#L93)):*
> ```JavaScript
>         return {};
>       }
>       const choice = js.results[Math.floor(Math.random() * js.results.length)];
> ```

TODO: This function should not discard unused results, so that if the first server is not available to run a client measurement, that the client can quickly try the next server.
___
*[src/ndt7-download-worker.js, line 11 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIzpfQD4Dtc3zqYLSuo:-MIzpfQE0nztgDBE93NB:b-7ts7va) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/src/ndt7-download-worker.js#L11)):*
> ```JavaScript
> const workerMain = function(ev) {
>   'use strict';
>   // TODO figure out where to put the secure/insecure choice
> ```

TODO: this is critical.
___
*[src/ndt7-upload-worker.js, line 10 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIzpiRP8SHsau89mzvL:-MIzpiRP8SHsau89mzvM:b-7ts7va) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/src/ndt7-upload-worker.js#L10)):*
> ```JavaScript
> // WebWorker that runs the ndt7 upload test
> const workerMain = function(ev) {
>   // TODO: Figure out the right place to choose wss vs ws
> ```

TODO: this is critical.
___
*[src/ndt7-upload-worker.js, line 114 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIzom_R35ckgzCW0nO6:-MIzom_R35ckgzCW0nO7:b-sp2m9t) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/src/ndt7-upload-worker.js#L114)):*
> ```JavaScript
>     if (t >= previous + clientMeasurementInterval) {
>       const numBytes = total - sock.bufferedAmount;
>       // ms * seconds / ms = seconds
> ```

nit: comment appears out of sync with logic.

**Review by stephen-soltesz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/ndt7-js/2)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mattcat10 and @pboothe)

___
*[README.md, line 2 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu25bN2-5H4u7AHBuU:-MIzq7hw9-Rn0PWIdp5D:b-896fix) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/README.md#L2)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

nit: word choice.
</blockquote></details>

Done.
___
*[README.md, line 5 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu2C7RCKQGDaIEhkKf:-MIzq8PxDNmU1w4mIPEP:b-896fix) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/README.md#L5)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

nit: in the examples/ dir.
</blockquote></details>

Done.
___
*[examples/node-client.js, line 22 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIu3oAl7PmOOi7-rF0h:-MIzqEZG1wt9g1Bi9OGE:b-896fix) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/examples/node-client.js#L22)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

nit: comment is out of sync with logic.
</blockquote></details>

Done.
___
*[src/ndt7-upload-worker.js, line 114 at r2](https://reviewable.io/reviews/m-lab/ndt7-js/2#-MIzom_R35ckgzCW0nO6:-MIzs25S-fM3soWkKpy2:b-896fix) ([raw file](https://github.com/m-lab/ndt7-js/blob/7fbbed0ca7b3c80ff7722c3b05eb6a1e7a8d5fa4/src/ndt7-upload-worker.js#L114)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

nit: comment appears out of sync with logic.
</blockquote></details>

Done.

**Review by stephen-soltesz — APPROVED** [review]
:lgtm: - with the following caveats:

* Unit tests are incomplete
* Locate API results are not saved and reused if the ndt server declines to run a measurement
* Upload rates are not calculated using actual runtime
* Selection of the non-TLS websocket protocol is hard coded.

> *[Reviewable](https://reviewable.io/reviews/m-lab/ndt7-js/2#-:-MIztBQh4ETRdYS5qlo2:bpbw5hx)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @mattcat10 and @pboothe)

**Pull request merged**